### PR TITLE
New resource: `azurerm_sentinel_alert_rule_nrt`

### PR DESCRIPTION
**Pull request by magodo**
This PR adds a new resource `azurerm_sentinel_alert_rule_nrt`, which is another alert rule variant. Besides, it adds a new block `nrt_template` in the data source `azurerm_sentinel_alert_rule_template` to allow users to retrieve NRT template, which can be in turn used to create the nrt alert rule.

## Test

```shell
💢  TF_ACC=1 go test -v -timeout=20h ./internal/services/sentinel  -run='TestAccSentinelAlertRuleTemplateDataSource_'
=== RUN   TestAccSentinelAlertRuleTemplateDataSource_fusion
=== PAUSE TestAccSentinelAlertRuleTemplateDataSource_fusion
=== RUN   TestAccSentinelAlertRuleTemplateDataSource_securityIncident
=== PAUSE TestAccSentinelAlertRuleTemplateDataSource_securityIncident
=== RUN   TestAccSentinelAlertRuleTemplateDataSource_scheduled
=== PAUSE TestAccSentinelAlertRuleTemplateDataSource_scheduled
=== RUN   TestAccSentinelAlertRuleTemplateDataSource_nrt
=== PAUSE TestAccSentinelAlertRuleTemplateDataSource_nrt
=== CONT  TestAccSentinelAlertRuleTemplateDataSource_fusion
=== CONT  TestAccSentinelAlertRuleTemplateDataSource_scheduled
=== CONT  TestAccSentinelAlertRuleTemplateDataSource_nrt
=== CONT  TestAccSentinelAlertRuleTemplateDataSource_securityIncident
--- PASS: TestAccSentinelAlertRuleTemplateDataSource_scheduled (307.00s)
--- PASS: TestAccSentinelAlertRuleTemplateDataSource_securityIncident (325.58s)
--- PASS: TestAccSentinelAlertRuleTemplateDataSource_fusion (415.01s)
--- PASS: TestAccSentinelAlertRuleTemplateDataSource_nrt (475.69s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/sentinel      475.707s


💤  TF_ACC=1 go test -v -timeout=20h ./internal/services/sentinel  -run='TestAccSentinelAlertRuleNrt_'
=== RUN   TestAccSentinelAlertRuleNrt_basic
=== PAUSE TestAccSentinelAlertRuleNrt_basic
=== RUN   TestAccSentinelAlertRuleNrt_complete
=== PAUSE TestAccSentinelAlertRuleNrt_complete
=== RUN   TestAccSentinelAlertRuleNrt_update
=== PAUSE TestAccSentinelAlertRuleNrt_update
=== RUN   TestAccSentinelAlertRuleNrt_requiresImport
=== PAUSE TestAccSentinelAlertRuleNrt_requiresImport
=== RUN   TestAccSentinelAlertRuleNrt_withAlertRuleTemplateGuid
=== PAUSE TestAccSentinelAlertRuleNrt_withAlertRuleTemplateGuid
=== CONT  TestAccSentinelAlertRuleNrt_basic
=== CONT  TestAccSentinelAlertRuleNrt_requiresImport
=== CONT  TestAccSentinelAlertRuleNrt_update
=== CONT  TestAccSentinelAlertRuleNrt_complete
=== CONT  TestAccSentinelAlertRuleNrt_withAlertRuleTemplateGuid
--- PASS: TestAccSentinelAlertRuleNrt_update (232.63s)
--- PASS: TestAccSentinelAlertRuleNrt_requiresImport (835.64s)
--- PASS: TestAccSentinelAlertRuleNrt_complete (899.41s)
--- PASS: TestAccSentinelAlertRuleNrt_withAlertRuleTemplateGuid (1024.94s)
--- PASS: TestAccSentinelAlertRuleNrt_basic (1420.47s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/sentinel      1420.488s
```

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@magodo what's "NRT" mean here?

**Comment by magodo** [comment]
Means "Near Real Time", I've documented it in the resource doc. For more info, please check: https://docs.microsoft.com/en-us/azure/sentinel/create-nrt-rules

**Comment by kaovd** [comment]
Thanks for this was looking at opening this one at some point (along with ThreatIntelligence) but see you got to it first 👍

**Comment by magodo** [comment]
@kaovd THe TI rule will be the next ;)

**Comment by kaovd** [comment]
Any updates on getting approval for this one @magodo  ?

**Comment by magodo** [comment]
Kindly ping @tombuildsstuff, is there anything else I can do to make this merged?

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+						string(securityinsight.AttackTacticInitialAccess),
+						string(securityinsight.AttackTacticLateralMovement),
+						string(securityinsight.AttackTacticPersistence),
+						string(securityinsight.AttackTacticPrivilegeEscalation),
+						string(securityinsight.AttackTacticPreAttack),
+					}, false),
+				},
+			},
+
+			"incident_configuration": {
```
Comment (katbyte) [inline internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:123]: configuration is redundant
```suggestion
			"incident": {
```

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+				Type:     pluginsdk.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				MinItems: 1,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"create_incident": {
```
Comment (katbyte) [inline internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:131]: boolean
```suggestion
						"create_incident_enabled": {
```

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+										Elem: &pluginsdk.Schema{
+											Type: pluginsdk.TypeString,
+											ValidateFunc: validation.StringInSlice([]string{
+												string(securityinsight.AlertDetailDisplayName),
+												string(securityinsight.AlertDetailSeverity),
+											},
+												false),
+										},
+									},
+									"group_by_custom_details": {
```
Comment (katbyte) [inline internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:188]: group is in the block name?
```suggestion
									"by_entities": {
										Type:     pluginsdk.TypeList,
										Optional: true,
										Elem: &pluginsdk.Schema{
											Type:         pluginsdk.TypeString,
											ValidateFunc: validation.StringInSlice(entityMappingTypes, false),
										},
									},
									"by_alert_details": {
										Type:     pluginsdk.TypeList,
										Optional: true,
										Elem: &pluginsdk.Schema{
											Type: pluginsdk.TypeString,
											ValidateFunc: validation.StringInSlice([]string{
												string(securityinsight.AlertDetailDisplayName),
												string(securityinsight.AlertDetailSeverity),
											},
												false),
										},
									},
									"by_custom_details": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Some comments left inline @magodo

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go (hunk 1 above)
Comment (magodo) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:123]: This name is already used by the scheduled alert rule: https://github.com/hashicorp/terraform-provider-azurerm/blob/26fcf8993da3e2f5d06126ec7d3f4759492028b6/internal/services/sentinel/sentinel_alert_rule_scheduled_resource.go#L142.

Shall we keep it consistent and add a "TODO-4.0" for this?

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go (hunk 2 above)
Comment (magodo) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:131]: Same name This name is already used by the scheduled alert rule: https://github.com/hashicorp/terraform-provider-azurerm/blob/26fcf8993da3e2f5d06126ec7d3f4759492028b6/internal/services/sentinel/sentinel_alert_rule_scheduled_resource.go#L150.

Shall we keep it consistent and add a "TODO-4.0" for this?

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go (hunk 3 above)
Comment (magodo) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:188]: Same name This name is already used by the scheduled alert rule: https://github.com/hashicorp/terraform-provider-azurerm/blob/26fcf8993da3e2f5d06126ec7d3f4759492028b6/internal/services/sentinel/sentinel_alert_rule_scheduled_resource.go#L187.

Shall we keep it consistent and add a "TODO-4.0" for this?

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go (hunk 1 above)
Comment (katbyte) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:123]: lets use the better name and add a todo, thanks

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go (hunk 2 above)
Comment (katbyte) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:131]: lets use the better name and add a todo, thanks

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go (hunk 3 above)
Comment (katbyte) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:188]: lets use the better name and add a todo, thanks

**Comment by magodo** [comment]
@katbyte I've made the changes per your comments. Whilst I'm not sure whether the naming change really worth against the ugly code...

File: internal/services/sentinel/sentinel_alert_rule_nrt_resource.go [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+										Optional: true,
+										Default:  true,
+									},
+									"lookback_duration": {
+										Type:         pluginsdk.TypeString,
+										Optional:     true,
+										ValidateFunc: validate.ISO8601Duration,
+										Default:      "PT5M",
+									},
+									"reopen_closed_incidents": {
```
Comment (katbyte) [inline internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:153]: sorry realized that the block name is `grouping` not `group_by` - which without changing the block name ti make the code better they should be
```suggestion
									"by_lookback_duration": {
										Type:         pluginsdk.TypeString,
										Optional:     true,
										ValidateFunc: validate.ISO8601Duration,
										Default:      "PT5M",
									},
									"by_reopen_closed_incidents": {
```
so its grouping.by_x

Comment (magodo) [reply internal/services/sentinel/sentinel_alert_rule_nrt_resource.go:153]: @katbyte Presumably you mean the `entities`, `alert_details` and `custom_details`. I've made the change then.

**Review by katbyte — APPROVED** [review]
Looks good aside from tests filing:
image

**Comment by magodo** [comment]
@katbyte My bad.. Updated the test to reflect the renaming change now:

```shell
💤  TF_ACC=1 go test -v -timeout=20h ./internal/services/sentinel -run='TestAccSentinelAlertRuleNrt_complete'
=== RUN   TestAccSentinelAlertRuleNrt_complete
=== PAUSE TestAccSentinelAlertRuleNrt_complete
=== CONT  TestAccSentinelAlertRuleNrt_complete
--- PASS: TestAccSentinelAlertRuleNrt_complete (160.92s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/sentinel      160.926s
```

**Review by katbyte — APPROVED** [review]

**Pull request merged**